src/Sanitize.php

```diff
@@ -639,6 +687,68 @@ public function do_strip_htmltags(array $match)
         }
     }
 
+    /**
+     * Keep only allowed HTML elements (tags) and their allowed attributes.
+     * @return void
+     */
+    protected function enforce_allowed_html_nodes(\DOMNode $element, bool $allow_data_attr = true, bool $allow_aria_attr = true)
+    {
+        if ($element instanceof \DOMElement) {
+            $tag = $element->tagName;
+            $div = null;
+            $doc = $element->ownerDocument;
+            if ($doc !== null) {
+                $div = $doc->getElementsByTagName('body')->item(0);
+                if ($div !== null) {
+                    $div = $div->firstChild;
+                    if ($div !== null && $div->nodeName !== 'div') {
+                        $div = null;
+                    }
+                }
+            }
+            if ($element !== $div
+                && !in_array($tag, ['html', 'head', 'body'])
+                && !isset($this->allowed_html_elements_with_attributes[$tag])) {
+                if (!in_array($tag, ['script', 'style'])) {
+                    // Preserve children inside the disallowed element
+                    for ($i = $element->childNodes->length - 1; $i >= 0; $i--) {
+                        $child = $element->childNodes->item($i);
+                        if ($child === null) {
+                            continue;
+                        }
+                        if ($child instanceof \DOMText) {
+                            $child->nodeValue = htmlspecialchars($child->nodeValue ?? '', ENT_QUOTES, 'UTF-8');
+                        }
+                        $element->before($child);
+                        $this->enforce_allowed_html_nodes($child, $allow_data_attr, $allow_aria_attr);
+                    }
+                }
+                $element->remove();
+            }
+            $allowed_attrs = array_merge($this->allowed_html_elements_with_attributes[$tag] ?? [], $this->allowed_html_attributes);
+            for ($i = $element->attributes->length - 1; $i >= 0; $i--) {
+                $attr = $element->attributes[$i]->nodeName;
+                // Skip data-*, aria-* if allowed
+                if (($allow_data_attr && str_starts_with($attr, 'data-'))
+                    || ($allow_aria_attr && str_starts_with($attr, 'aria-'))) {
+                    continue;
+                }
+
+                if (!in_array($attr, $allowed_attrs, true)) {
+                    $element->removeAttributeNS(null, $attr);
```

Comment:
That looks a bit fishy. Consider something like that instead:

```suggestion
                    $element->removeChild($element->attributes[$i]);
```

Reply:
That doesn't work 🤔

Reply:
My bad. What about something like:

```suggestion
                    $element->removeAttributeNode($element->attributes[$i]);
```

Reply:
That works and does the same thing as expected now